File: katas/5 kyu/best-travel/solution.php
<!-- best-travel -->
<?php
use PHPUnit\Framework\TestCase;
/* ------------------------------- MY SOLUTION ------------------------------ */
function chooseBestSum($t, $k, $ls) {
  // the $arr part fills $arr with all possible combinations
  $arr = array(array( ));

  foreach ($ls as $element)
    foreach ($arr as $combination)
      array_push($arr, array_merge(array($element), $combination));

  $longest = [];

  foreach($arr as $subar) {
    if (count($subar) != $k)
      continue;
    
    if (array_sum($subar) > array_sum($longest) and array_sum($subar) <= $t)
      $longest = $subar;
  }

  return count($longest) > 0 ? array_sum($longest) : null;
}
/* ----------------------------- CLEVER SOLUTION ---------------------------- */
/* ---------------------------------- TEST ---------------------------------- */
class BestTravelTestCases extends TestCase {
  private function revTest($actual, $expected) {
      $this->assertSame($expected, $actual);
  }
  public function testCountOnesBasics() {
      $ts = [50, 55, 56, 57, 58];
      $this->revTest(chooseBestSum(163, 3, $ts), 163);
      $ts = [50];
      $this->revTest(chooseBestSum(163, 3, $ts), null);
      $ts = [91, 74, 73, 85, 73, 81, 87];
      $this->revTest(chooseBestSum(230, 3, $ts), 228);
  }
}

File: katas/5 kyu/best-travel/solution.py
# best-travel
from tkinter.messagebox import NO
import codewars_test as test
# -------------------------------- MY SOLUTION ------------------------------- #
from itertools import combinations
def choose_best_sum(t, k, ls):
  arr = combinations(ls , k)
  arr_sum = [sum(x) for x in arr if sum(x) <= t]
  return max(arr_sum) if arr_sum != [] else None
# ------------------------------ CLEVER SOLUTION ----------------------------- #
# ----------------------------------- TEST ----------------------------------- #

test.it("Bigger numbers")
xs = [100, 76, 56, 44, 89, 73, 68, 56, 64, 123, 2333, 144, 50, 132, 123, 34, 89]
test.assert_equals(choose_best_sum(230, 4, xs), 230)
test.assert_equals(choose_best_sum(430, 5, xs), 430)
test.assert_equals(choose_best_sum(430, 8, xs), None)
